osu.Game.Rulesets.Mania/Beatmaps/StageDefinition.cs

```diff
@@ -16,12 +16,40 @@ public struct StageDefinition
         /// </summary>
         public int Columns;
 
+        /// <summary>
+        /// Whether the SpecialStyle flag is set.
+        /// </summary>
+        public bool SpecialStyle;
+
+        /// <summary>
+        /// Whether this is a special style map.
+        /// </summary>
+        /// <remarks>
+        /// Aside from 6K and 8K, special style is almost never used in other
+        /// keymodes, so this function always gives false for non-6/8K modes.
+        /// This also matches stable behavior of only allowing configuration
+        /// of special style in 6/8K.
+        /// </remarks>
+        public bool UseSpecialStyle()
+        {
+            if (Columns != 6 && Columns != 8)
+                return false;
+
+            return SpecialStyle;
+        }
+
         /// <summary>
         /// Whether the column index is a special column for this stage.
         /// </summary>
         /// <param name="column">The 0-based column index.</param>
         /// <returns>Whether the column is a special column.</returns>
-        public bool IsSpecialColumn(int column) => Columns % 2 == 1 && column == Columns / 2;
+        public bool IsSpecialColumn(int column)
+        {
+            if (UseSpecialStyle())
+                return column == 0;
```

Comment:
I'm very much not a fan of magic conditionals like this, relying on the numerical index of a column to determine if it's special. It's very flaky logic, and will get even more flaky once switching the special column position is reintroduced (which I understand isn't supported at this stage of the PR?)

I'm thinking `StageDefinition` needs an overhaul. The key count and special flags should be passed via ctor to sanely initialise the stage definition. The ctor would then assign types to the columns, and this would reduce to a simple read of that type.

Reply:
I think the user skin setting needs to be implemented before I can imagine the future direction of this, since that will completely change this implementation.